test/e2e/basic_test.go

```diff
@@ -561,6 +559,115 @@ func testReconcileGatewayRoutesOnStartup(t *testing.T, data *TestData, isIPv6 bo
+	createAndWaitForPod(t, data, data.createNginxPodOnNode, "test-clean-stale-route-pod", nodeName(0), testNamespace, false)
+
+	if len(clusterInfo.podV4NetworkCIDR) != 0 {
+		t.Logf("Running IPv4 test")
+		testCleanStaleClusterIPRoutes(t, data, false)
+	}
+	if len(clusterInfo.podV6NetworkCIDR) != 0 {
+		t.Logf("Running IPv6 test")
+		testCleanStaleClusterIPRoutes(t, data, true)
+	}
+}
+
+func testCleanStaleClusterIPRoutes(t *testing.T, data *TestData, isIPv6 bool) {
+	ipProtocol := corev1.IPv4Protocol
+	virtualIP := config.VirtualServiceIPv4
+	if isIPv6 {
+		ipProtocol = corev1.IPv6Protocol
+		virtualIP = config.VirtualServiceIPv6
+	}
+	// Create two test ClusterIPs.
+	svc, err := data.createNginxClusterIPService(fmt.Sprintf("test-clean-stale-route-svc1-%v", isIPv6), testNamespace, false, &ipProtocol)
+	require.NoError(t, err)
+	require.NotEqual(t, "", svc.Spec.ClusterIP, "ClusterIP should not be empty")
+	svc, err = data.createNginxClusterIPService(fmt.Sprintf("test-clean-stale-route-svc2-%v", isIPv6), testNamespace, false, &ipProtocol)
+	require.NoError(t, err)
+	require.NotEqual(t, "", svc.Spec.ClusterIP, "ClusterIP should not be empty")
+	time.Sleep(time.Second)
+
+	node := nodeName(0)
+	pod := getAntreaPodName(t, data, node)
+	restartAntreaAgentPod(t, data, pod, node)
+	defer func() {
+		// "cleanup": delete agent to ensure the restart count goes back to 0
+		// this will also take care of deleting the flow in case of test failure
+		if _, err := data.deleteAntreaAgentOnNode(node, 30 /* grace period in seconds */, defaultTimeout); err != nil {
+			t.Logf("Error when restarting antrea-agent on Node '%s': %v", node, err)
+		}
+	}()
+
+	antreaGWName, err := data.GetGatewayInterfaceName(antreaNamespace)
+	if err != nil {
+		t.Fatalf("Failed to detect gateway interface name from ConfigMap: %v", err)
+	}
+	getServiceGatewayRoutes := func() (routes []*net.IPNet, err error) {
+		var cmd []string
+		mask := 32
+		if !isIPv6 {
+			cmd = []string{"ip", "route", "list", "dev", antreaGWName}
+		} else {
+			cmd = []string{"ip", "-6", "route", "list", "dev", antreaGWName}
+			mask = 128
+		}
+		stdout, stderr, err := data.runCommandFromPod(antreaNamespace, pod, agentContainerName, cmd)
+		if err != nil {
+			return nil, fmt.Errorf("error when running ip command in Pod '%s': %v - stdout: %s - stderr: %s", pod, err, stdout, stderr)
+		}
+		re := regexp.MustCompile(fmt.Sprintf(`([^\s]+) via %s`, virtualIP))
+		for _, line := range strings.Split(stdout, "\n") {
+			var err error
+			matches := re.FindStringSubmatch(line)
+			if len(matches) == 0 {
+				continue
+			}
+			if net.ParseIP(matches[1]) != nil {
+				matches[1] = fmt.Sprintf("%s/%d", matches[1], mask)
+			}
+			var route *net.IPNet
+			if _, route, err = net.ParseCIDR(matches[1]); err != nil {
+				return nil, fmt.Errorf("%s is not a valid net CIDR", matches[1])
+			}
+			routes = append(routes, route)
+		}
```

Comment:
do you think we can unify the code that reads and parse routes with `testReconcileGatewayRoutesOnStartup`?

Reply:
Done.